Add IEdmTargetPath docs (#313)
MicrosoftDocs/OData-docs: Odata-docs/odatalib/v7/edm/iedmtargetpath-for-annotations-target.md

---

> title: "Use IEdmTargetPath for out-of-line annotations target"
> description: "Use IEdmTargetPath for out-of-line annotations target"

Reviewer: ```suggestion
title: "Set target for out-of-line annotations"
description: "Set target for out-of-line annotations"
```

author: kenitoinc
ms.author: kemunga
ms.date: 05/2/2023
ms.topic: article
 
---

> # Use IEdmTargetPath for out-of-line annotations target

Reviewer: ```suggestion
# Set target for out-of-line annotations
```

**Applies To**: [!INCLUDE[appliesto-odataclient](../../../includes/appliesto-odatalib-v7.md)]

The Edm library lets you [define annotations](/odata/odatalib/edm/define-annotations).

> [This](/odata/odatalib/edm/set-annotations-schema) documentation describes how to define **out-of-line** annotations that targets a model element.

Reviewer: The Edm library lets you [define annotations](/odata/odatalib/edm/define-annotations). When annotations relevant to a particular model element are expressed **out-of-line**/**externally** in an `Annotations` element, a path expression that uniquely identifies the target element must be specified.

This page demonstrates how to set the target for out-of-line annotations.

> ## Background

Reviewer: "Background" is okay but it could also be "Introduction" or a any other term that communicates that the section illustrates the concept of out-of-line annotations

> 

Reviewer 2: At least, 
Why do customers need to annotate on the path? "For example, customer can annotate the path using different permissions, bla bla"

How can customers to annotate on the path? 
1) We introduce "IEdmTargetPath", what is it? how to build it? 
2) How can customers use this interface to annotate the path?
3) As example, here's the sample codes to use this interface for 'writing'
4) As example, here's the sample codes to use this interface for 'reading'
5) Advanced, maybe we can have some sample codes to illustrate the usage on Uri parser (if apply)

> 

Author: Sample codes for reading and writing will be in the sections below, not the background section.

> Below is a simple Edm model with an out-of-line annotation that targets a model element.

Reviewer: Consider the follow metadata describing a service where a model element `Employee` defined in the `Ns` namespace has annotations defined out-of-line in the `Default` namespace:

```xml
<?xml version="1.0" encoding="utf-8"?>
<edmx:Edmx Version="4.0" xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx">
    <edmx:DataServices>
        <!-- Model elements -->
        <Schema Namespace="Default" xmlns="http://docs.oasis-open.org/odata/ns/edm">
            <EntityContainer Name="Container">
                <EntitySet Name="Employees" EntityType="Ns.Employee" />
            </EntityContainer>
            <Annotations Target="Ns.Employee">
                <Annotation Term="Ns.MyTerm" String="Name OutOfLine MyTerm Value" />
            </Annotations>

>         </Schema>

Reviewer: You want this sample to demonstrate both the model element together with the out-of-line annotations targeting that element. Please include the element in the `Ns` namespace

    </edmx:DataServices>
</edmx:Edmx>
```

> 

Reviewer: Here you should explain what you mean to set the target...
```suggestion
Take note of the "Target" attribute on the `Annotations` element. It specifies the path expression to the target model element, in this case the fully qualified name of the element comprising of the namespace and the element name.
```

> `IEdmTargetPath` allows us to define out-of-line annotations that target a path to a model element.

Reviewer 2: add a linker to 'IEdmTargetPath' definition (source)

Reviewer 3: Are our reference docs up to date. I think it would be better to link to the reference documentation if it's available, otherwise you can link to the source.

> `IEdmTargetPath` allows us to define out-of-line annotations that target a path to a model element.

Reviewer: I think this line should come after the examples below

> Below are examples of paths to a model element.

Reviewer: You want to introduce what the feature you implemented will help someone achieve:
```suggestion
The path expression must resolve to a model element in the entity model which means it can comprise of multiple segments. The path expression may be:
- Qualified name of a schema child - e.g., `Ns.Employee` from the metadata above
- Qualified name of a schema child following by a forward slash and name of child element
- Qualified name of a structured type followed by zero or more property, navigation property, or type cast segments, each segment starting with a forward slash - e.g., `Ns.Employee/Name` from the sample metadata above.
- Qualified name of an entity container followed by a segment containing a singleton or entity set name and zero or more property, navigation property, or type segments - e.g., `Default.Container/Employees/Name` from the sample metadata above. 

Below are examples of path expressions to model elements.
```

Author: Paths to a model always begin from the Entity container

https://docs.oasis-open.org/odata/odata-csdl-xml/v4.01/odata-csdl-xml-v4.01.html#sec_TargetPath

```
MySchema.MyEntityContainer/MyEntitySet/MyProperty
MySchema.MyEntityContainer/MyEntitySet/MyNavigationProperty
MySchema.MyEntityContainer/MyEntitySet/MySchema.MyEntityType/MyProperty
MySchema.MyEntityContainer/MyEntitySet/MySchema.MyEntityType/MyNavProperty
MySchema.MyEntityContainer/MyEntitySet/MyComplexProperty/MyProperty
MySchema.MyEntityContainer/MyEntitySet/MyComplexProperty/MyNavigationProperty
MySchema.MyEntityContainer/MySingleton/MyComplexProperty/MyNavigationProperty
```

> Below is a simple Edm model with an out-of-line annotation that targets a path to a model element.
> 
> ```xml
> <?xml version="1.0" encoding="utf-8"?>
> <edmx:Edmx Version="4.0" xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx">
>     <edmx:DataServices>
>         <!-- Model elements -->
>         <Schema Namespace="Default" xmlns="http://docs.oasis-open.org/odata/ns/edm">
>             <EntityContainer Name="Container">
>                 <EntitySet Name="Employees" EntityType="Ns.Employee" />
>             </EntityContainer>
>             <Annotations Target="Default.Container/Employees/Ns.Manager/EmployeeId">
>                 <Annotation Term="Ns.MyTerm" String="Name OutOfLine MyTerm Value" />
>             </Annotations>
>         </Schema>
>     </edmx:DataServices>
> </edmx:Edmx>
> ```

Reviewer: How is this different from the earlier sample and what new value does it add?
I think you should get rid of it...
Instead you should introduce the following sections where you're going to describe how to use your feature to set the target
```suggestion
In the sections that follow, we cover how to set the target for an out-of-line annotation using the Edm model builder.
```

## Create a .NET Core console application
Create a .NET Core console application and import the `Microsoft.OData.Edm` nuget package:

---

# [Visual Studio](#tab/visual-studio)

In the Visual Studio **Package Manager Console**:
```powershell
Install-Package Microsoft.OData.Edm
```

# [.NET Core CLI](#tab/netcore-cli)

```dotnetcli
dotnet add package Microsoft.OData.Edm
```

---

> ## Configure the Edm model

Reviewer: ```suggestion
## Define the Edm model
```

> Add a _SampleModel.cs_ and replace the code with the following code

Reviewer: ```suggestion
Add a file named _SampleModel.cs_ and replace the contents with the following code:
```

```csharp
using Microsoft.OData.Edm;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AnnotationsTarget
{

>     internal class SampleModel

Reviewer 2: SampleModel is un-necessary here?

Maybe just keep the 'GetEdmModel' or 'CreateEdmModel' method for simplicity

    {

>         public SampleModel()

Reviewer: In my opinion, the only element that you need to add in your model for purposes of what you're trying to illustrate is `Customer` and the `Customers` entity set. You don't need everything else. It just makes your sample unnecessarily long, distracting and full of noise.

        {
            this.Model = new EdmModel();
            this.Address = new EdmComplexType("NS", "Address");
            this.Address.AddStructuralProperty("Road", EdmCoreModel.Instance.GetString(false));
            this.Model.AddElement(this.Address);

>             this.DerivedAddress = new EdmComplexType("NS", "DerivedAddress", this.Address);

Reviewer 2: do we need all of these 'elements' in the sample?

                new EdmNavigationPropertyInfo()
                {
                    Name = "City",
                    Target = this.City,
                    TargetMultiplicity = EdmMultiplicity.One,
                });

            this.NavUnderCustomer = this.Customer.AddUnidirectionalNavigation(
                new EdmNavigationPropertyInfo()
                {
                    Name = "Cities",
                    Target = this.City,
                    TargetMultiplicity = EdmMultiplicity.Many,
                });

            this.NameProperty = this.Customer.DeclaredProperties.Where(x => x.Name == "Name").FirstOrDefault();
            this.AddressProperty = this.Customer.DeclaredProperties.Where(x => x.Name == "Address").FirstOrDefault();

            this.Container = new EdmEntityContainer("NS", "Default");
            this.EntitySet = new EdmEntitySet(this.Container, "Customers", this.Customer);
            this.Singleton = new EdmSingleton(this.Container, "Me", this.Customer);
            this.Container.AddElement(this.EntitySet);
            this.Container.AddElement(this.Singleton);

            this.Model.AddElement(this.Container);
        }

>         public EdmModel Model { get; private set; }

Reviewer 2: all of these properties are un-necessary in the doc

        public EdmEntityContainer Container { get; private set; }
        public EdmEntitySet EntitySet { get; private set; }
        public EdmSingleton Singleton { get; private set; }
        public EdmEntityType Customer { get; private set; }
        public EdmEntityType VipCustomer { get; private set; }
        public EdmEntityType City { get; private set; }
        public EdmComplexType Address { get; private set; }
        public EdmComplexType DerivedAddress { get; private set; }
        public IEdmProperty NameProperty { get; private set; }
        public IEdmProperty AddressProperty { get; private set; }
        public EdmNavigationProperty NavUnderComplex { get; private set; }
        public EdmNavigationProperty NavUnderCustomer { get; private set; }
    }
}

```

## Write out-of-line annotations

> Replace the contents of _Program.cs_ file with the following code:

Reviewer 2: At least sentences to illustrate what will do in this session? 
Or what you want to achieve? Use a 'concrete' scenario could be better.

for example:  I want to limit the 'user' access for 'customers/{key}/detailInfo'? 
How can we use the annotations on this path to specify the 'allowed' user accesses?

```csharp
using Microsoft.OData.Edm.Csdl;
using Microsoft.OData.Edm.Vocabularies;
using Microsoft.OData.Edm;
using System.Reflection;
using System.ComponentModel;
using Microsoft.OData.Edm.Validation;
using System.Xml;
using static System.Runtime.InteropServices.JavaScript.JSType;
using System.Xml.Linq;

namespace AnnotationsTarget
{
    internal class Program
    {

>         static void Main(string[] args)

Reviewer: You could merge this code snippet with the previous one and have 3 methods:
- InitializeModel()
- AddOutOfLineAnnotationsTargetingAModelElement() // Or a better method name
- SerializeModel()

I think they'd amply demonstrate what you wanted to

        {

>             SampleModel sampleModel = new SampleModel();

Reviewer 2: For demo, maybe re-organize the 'main' body?
Create
1) one method to play building the annotation on target path during writing
3) one method to play consume the annotation on target path during reading.

            model.AddElement(term);
            EdmVocabularyAnnotation annotation = new EdmVocabularyAnnotation(targetPath, term, new EdmStringConstant("Name OutOfLine MyTerm Value"));
            annotation.SetSerializationLocation(model, EdmVocabularyAnnotationSerializationLocation.OutOfLine);
            model.SetVocabularyAnnotation(annotation);

            WriteXml(model);
        }

        private static void WriteXml(IEdmModel model, CsdlTarget target = CsdlTarget.OData)
        {
            using (StringWriter sw = new StringWriter())
            {
                XmlWriterSettings settings = new XmlWriterSettings();
                settings.Encoding = System.Text.Encoding.UTF8;
                settings.Indent = true;

                using (XmlWriter xw = XmlWriter.Create(sw, settings))
                {
                    IEnumerable<EdmError> errors;
                    CsdlWriter.TryWriteCsdl(model, xw, target, out errors);
                    xw.Flush();
                }

                string output = sw.ToString();
                Console.Write(output);
            }
        }
    }
}

> ```

Reviewer: Before inviting the reader to run the application, you want to explain your code sample
- Describe what is happening in `InitializeModel` method. In brief since it's not the focus of your doc, e.g., In the `InitializeModel` method we're initializing the model... yada yada
- Describe what is happening in "AddOutOf..." method. In detail since that's the focus of your doc.
- Describe in 1 or 2 sentences what is happening in `SerializeModel` method.

> Run the console application.

Reviewer: If the preceding header is a level 2, you should make this a level 3

      <ComplexType Name="DerivedAddress" BaseType="NS.Address">
        <Property Name="MyRoad" Type="Edm.String" Nullable="false" />
      </ComplexType>
      <EntityType Name="Customer">
        <Key>
          <PropertyRef Name="Id" />
        </Key>
        <Property Name="Id" Type="Edm.Int32" Nullable="false" />
        <Property Name="Name" Type="Edm.String" Nullable="false" />
        <Property Name="Address" Type="NS.Address" Nullable="false" />
        <NavigationProperty Name="Cities" Type="Collection(NS.City)" />
      </EntityType>
      <EntityType Name="VipCustomer" BaseType="NS.Customer" />
      <EntityType Name="City">
        <Property Name="Name" Type="Edm.String" Nullable="false" />
      </EntityType>
      <Term Name="MyTerm" Type="Edm.String" />
      <EntityContainer Name="Default">
        <EntitySet Name="Customers" EntityType="NS.Customer" />
        <Singleton Name="Me" Type="NS.Customer" />
      </EntityContainer>
      <Annotations Target="NS.Default/Customers/Name">
        <Annotation Term="NS.MyTerm" String="Name OutOfLine MyTerm Value" />
      </Annotations>
    </Schema>
  </edmx:DataServices>
</edmx:Edmx>
```

## Read out-of-line annotation

> Replace the contents of _Program.cs_ file with the following code:

Reviewer 2: sentences about what will you do in this session?

			<ComplexType Name=""Address"">
				<Property Name=""Road"" Type=""Edm.String"" Nullable=""false"" />
				<NavigationProperty Name=""City"" Type=""NS.City"" Nullable=""false"" />
			</ComplexType>
			<ComplexType Name=""DerivedAddress"" BaseType=""NS.Address"">
				<Property Name=""MyRoad"" Type=""Edm.String"" Nullable=""false"" />
			</ComplexType>
			<EntityType Name=""Customer"">
				<Key>
					<PropertyRef Name=""Id""/>
				</Key>
				<Property Name=""Id"" Type=""Edm.Int32"" Nullable=""false"" />
				<Property Name=""Name"" Type=""Edm.String"" Nullable=""false"" />
				<Property Name=""Address"" Type=""NS.Address"" Nullable=""false"" />
				<NavigationProperty Name=""Cities"" Type=""Collection(NS.City)"" />
			</EntityType>
			<EntityType Name=""VipCustomer"" BaseType=""NS.Customer"" />
			<EntityType Name=""City"">
				<Property Name=""Name"" Type=""Edm.String"" Nullable=""false"" />
			</EntityType>
			<Term Name=""MyTerm"" Type=""Edm.String"" />
			<EntityContainer Name=""Default"">
				<EntitySet Name=""Customers"" EntityType=""NS.Customer"" />
				<Singleton Name=""Me"" Type=""NS.Customer"" />
			</EntityContainer>
			<Annotations Target=""NS.Default/Customers/Name"">

> 				<Annotation Term=""NS.MyTerm"" String=""Name OutOfLine MyTerm Value"" />

Reviewer 2: 'Name OutOfLine MyTerm Value' for test codes is ok. but, in my opinion, as a 'doc' or 'tutorial', at least we should use 'meaningful' string for the value, or at least 'readable' string.

            IEnumerable<EdmError> errors;
            CsdlReader.TryParse(XElement.Parse(csdl).CreateReader(), out parsedModel, out errors);
            IEdmTargetPath tgtPath = parsedModel.GetTargetPath("NS.Default/Customers/Name");
            IEdmTerm fooBarTerm = parsedModel.FindDeclaredTerm("NS.MyTerm");
            string nameAnnotation = GetAnnotation(parsedModel, tgtPath, fooBarTerm, EdmVocabularyAnnotationSerializationLocation.OutOfLine);
            Console.WriteLine(nameAnnotation);
        }

        private static string GetAnnotation(IEdmModel model, IEdmVocabularyAnnotatable target, IEdmTerm term, EdmVocabularyAnnotationSerializationLocation location)
        {
            IEdmVocabularyAnnotation annotation = model.FindVocabularyAnnotations<IEdmVocabularyAnnotation>(target, term).FirstOrDefault();
            if (annotation != null)
            {
                IEdmStringConstantExpression stringConstant = annotation.Value as IEdmStringConstantExpression;
                if (stringConstant != null)
                {
                    return stringConstant.Value;
                }
            }

            return null;
        }
    }
}
```

> Run the console application.

Reviewer: If the preceding header is a level 2, you should make this a level 3. It might be confusing to have two "Run the console application" headers in the same document

The output on the console should be as below:

`Name OutOfLine MyTerm Value`

## Read all annotations with a specific annotation target

> We use the `Model.FindVocabularyAnnotations(IEdmTargetPath)` extension method.

Reviewer: Explain what you'll demonstrate in this section
